File: modules/expo-apple-intelligence/android/src/main/java/expo/modules/appleintelligence/ExpoAppleIntelligenceModule.kt
package expo.modules.appleintelligence

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoAppleIntelligenceModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoAppleIntelligence")

    AsyncFunction("queryAsync") { _: String ->
      throw UnsupportedOperationException(
        "Apple Intelligence querying is only available on iOS devices that support Apple Intelligence."
      )
    }

    AsyncFunction("requestHealthAuthorizationAsync") {
      throw UnsupportedOperationException("HealthKit is only available on iOS.")
    }

    AsyncFunction("getStepCountsAsync") { _: Int ->
      throw UnsupportedOperationException("HealthKit is only available on iOS.")
    }

    AsyncFunction("getHealthSummaryAsync") { _: Int ->
      throw UnsupportedOperationException("HealthKit is only available on iOS.")
    }

    AsyncFunction("requestCalendarAccessAsync") {
      throw UnsupportedOperationException("Calendar sync is only available on iOS.")
    }

    AsyncFunction("syncCalendarEventsAsync") { _: List<Map<String, String>> ->
      throw UnsupportedOperationException("Calendar sync is only available on iOS.")
    }

    AsyncFunction("removeCalendarEventsAsync") { _: List<String> ->
      throw UnsupportedOperationException("Calendar sync is only available on iOS.")
    }

    AsyncFunction("setStoredValueAsync") { _: String, _: String ->
      throw UnsupportedOperationException("Native storage is only available on iOS.")
    }

    AsyncFunction("getStoredValueAsync") { _: String ->
      throw UnsupportedOperationException("Native storage is only available on iOS.")
    }

    AsyncFunction("removeStoredValueAsync") { _: String ->
      throw UnsupportedOperationException("Native storage is only available on iOS.")
    }
  }
}

File: modules/expo-apple-intelligence/src/ExpoAppleIntelligenceModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import {
  CalendarWorkoutEvent,
  ExpoAppleIntelligenceModuleEvents,
  HealthSummary,
  QueryResult,
  StepCountRow,
} from './ExpoAppleIntelligence.types';

declare class ExpoAppleIntelligenceModule extends NativeModule<ExpoAppleIntelligenceModuleEvents> {
  queryAsync(prompt: string): Promise<QueryResult>;
  requestHealthAuthorizationAsync(): Promise<boolean>;
  getStepCountsAsync(days: number): Promise<StepCountRow[]>;
  getHealthSummaryAsync(days: number): Promise<HealthSummary>;
  requestCalendarAccessAsync(): Promise<boolean>;
  syncCalendarEventsAsync(events: CalendarWorkoutEvent[]): Promise<string[]>;
  removeCalendarEventsAsync(eventIds: string[]): Promise<number>;
  setStoredValueAsync(key: string, value: string): Promise<boolean>;
  getStoredValueAsync(key: string): Promise<string | null>;
  removeStoredValueAsync(key: string): Promise<boolean>;
}

export default requireNativeModule<ExpoAppleIntelligenceModule>('ExpoAppleIntelligence');
